packages/sources/stader-balance/src/endpoint/balance.ts

```diff
@@ -0,0 +1,741 @@
+          logger.error(errorMessage)
+          response = buildErrorResponse(errorMessage, providerDataRequestedUnixMs)
+        }
+
+        await this.responseCache.write(this.name, [
+          {
+            params: req,
+            response,
+          },
+        ])
+      }),
+    )
+  }
+
+  // Retrieve balances from the beacon chain for all validators in request
+  async queryValidatorBalances(
+    req: RequestParams,
+    settings: typeof config.settings,
+  ): Promise<ValidatorState[]> {
+    try {
+      const url = `/eth/v1/beacon/states/${req.stateId}/validators`
+      const statusList = req.validatorStatus?.join(',')
+      const batchSize = settings.BATCH_SIZE
+      const batchedAddresses = []
+      const responses: AxiosResponse<ProviderResponse>[] = []
+      const result: ValidatorState[] = []
+      const addresses = req.addresses
+      // Separate the address set into the specified batch size
+      // Add the batches as comma-separated lists to a new list used to make the requests
+      for (let i = 0; i < addresses.length / batchSize; i++) {
+        batchedAddresses.push(
+          addresses
+            .slice(i * batchSize, i * batchSize + batchSize)
+            .map(({ address }) => address)
+            .join(','),
+        )
+      }
+      // Make request to beacon API for every batch
+      // Break addresses down into groups to execute asynchronously
+      // Firing requests for all batches all at once could hit rate limiting for large address pools
+      const groupedBatches = chunkArray(batchedAddresses, settings.GROUP_SIZE)
+      for (const group of groupedBatches) {
+        await Promise.all(
+          group.map(async (address) => {
+            const options: AxiosRequestConfig = {
+              baseURL: settings.BEACON_RPC_URL,
+              url,
+              params: { id: address, status: statusList },
+            }
+            return axios.request<ProviderResponse>(options).then((response) => {
+              responses.push(response)
+            })
+          }),
+        )
+      }
+
+      // Flatten the results into single array for validators and balances
+      const validatorBatches = responses.map(({ data }) => data)
+      validatorBatches.forEach(({ data }) => {
+        data.forEach((validator) => {
+          result.push(validator)
+        })
+      })
+      return result
+    } catch (e) {
+      throw new Error('Failed to retrieve validator balances from Beacon chain')
+    }
+  }
+
+  // Perform validator level calculations for balance on each one
+  async performValidatorCalculations(
+    addresses: ValidatorAddress[],
+    validatorStateList: ValidatorState[],
+    balances: BalanceResponse[],
+    req: RequestParams,
+    validatorDeposit: BigNumber,
+    ethDepositContract: string,
+    commissionMap: Record<number, number>,
+    collateralEthMap: Record<number, BigNumber>,
+    blockTag: number,
+  ): Promise<void> {
```

Comment:
If we have more than ~3 parameters, these should be passed as an obj for readability

Reply:
(readonly, to avoid problems)

Reply:
I probably can't make it read-only because some of the parameters like `commissionMap` and `collateralEthMap` are updated on the fly and used in a couple different functions. But will move this to an object for sure